### DISCUSSION TITLE
SSL_READ fails forever with SSL_ERROR_WANT_READ

### DISCUSSION THREAD

**Discussion by Chris-IBM01** [openssl/openssl#25323] [Q&A]
I'm using SSL Library 3.0.9.0 over a non-blocking socket on Windows.  After the connection is established and the handshake is complete, I get an indication that there is data to be read on the socket.  I call SSL_read and it fails with SSL_ERROR_WANT_READ.  Repeated calls to SSL_read continue to fail forever with SSL_ERROR_WANT_READ.

Am I doing something wrong in response to this error, or is this a known bug?

**Comment by mattcaswell** [comment]
> I get an indication that there is data to be read on the socket.

Does it continue to indicate this after you have called `SSL_read`? Data available to read on the socket does not necessarily equate to application data available to read, e.g. it could be a protocol level message which doesn't result in any new application data. Calling SSL_read() should read pending data from the socket and process it. If no new application data is available after processing it, it will indicate SSL_ERROR_WANT_READ.

**Comment by asomers** [comment]
I have the same problem, albeit on FreeBSD instead of Windows.  After the remote side closes its connection, `SSL_read` returns `SSL_ERROR_WANT_READ` forever.  A the OS level, `poll` returns 1 to indicate that the socket is readable, and then `read` returns 0 to indicate EOF.  But `SSL_READ` still returns `SSL_ERROR_WANT_READ`.  `netstat` verifies that the remote side has closed the connection, because socket is in the `CLOSE_WAIT` state.
OpenSSL didn't always work this way.  The application works correctly with OpenSSL versions 3.0.16, 3.1.4, and 3.1.8.  But it fails with 3.2.0 and newer.
How is my application to tell that the socket is EOF?
